tree.go

```diff
@@ -100,16 +111,35 @@ func addDumpStatsToTree(tree treeprint.Tree, dumpStats *images.DumpStatsEntry) {
 }
 
 func addPsTreeToTree(tree treeprint.Tree, psTree *crit.PsTree) {
-	// processNodes is a recursive function to create
-	// a new branch for each process and add its child
-	// processes as child nodes of the branch.
+	/* processNodes is a recursive function to create
+	a new branch for each process and add its child
+	processes as child nodes of the branch. */
 	var processNodes func(treeprint.Tree, *crit.PsTree)
 	processNodes = func(tree treeprint.Tree, root *crit.PsTree) {
-		node := tree.AddMetaBranch(root.PId, root.Comm)
+		node := tree.AddMetaBranch(root.PID, root.Comm)
 		for _, child := range root.Children {
 			processNodes(node, child)
 		}
 	}
 	psTreeNode := tree.AddBranch("Process tree")
 	processNodes(psTreeNode, psTree)
 }
+
+func addFdsToTree(tree treeprint.Tree, fds []*crit.Fd) {
+	var node treeprint.Tree
+	for _, fd := range fds {
+		node = tree.FindByMeta(fd.PId)
+		if node == nil {
+			/* If the PID is not present in the process tree,
+			it is either invalid, or has been pruned while
+			building the process tree. In the first case,
+			the error will be handled by the process tree
+			builder. In the second case, we simply omit adding
+			those FDs. */
+			continue
+		}
```

Comment:
Hmm, in ExploreFds() we have the following code:

```Go
pID := process.GetPid()
// Get file with object IDs
idsImg, err := getImg(filepath.Join(c.inputDirPath, fmt.Sprintf("ids-%d.img", pID)), &criu_core.TaskKobjIdsEntry{})
if err != nil {
	return nil, err
}
filesID := idsImg.Entries[0].Message.(*criu_core.TaskKobjIdsEntry).GetFilesId()
// Get open file descriptors
fdInfoImg, err := getImg(filepath.Join(c.inputDirPath, fmt.Sprintf("fdinfo-%d.img", filesID)), &fdinfo.FdinfoEntry{})
if err != nil {
	return nil, err
}

fdEntry := Fd{PId: pID}
```

Why `tree.FindByMeta(fd.PId)` would return `nil`?

Reply:
This is twofold. Firstly, in case we have an invalid checkpoint, and still successfully make it to this stage, if `tree.FindByMeta()` does not find the node and returns nil, it will cause a panic due to nil pointer dereference. This can be handled more gracefully by checking for the nil value.
Secondly, this is related to the change in #87 - if we prune the process tree for a particular PID, the other PID nodes will not be present. Hence, for those PID's file descriptors, `tree.FindByMeta()` will return nil.

Reply:
>   in case we have an invalid checkpoint, and still successfully make it to this stage, if tree.FindByMeta() does not find the node and returns nil, it will cause a panic due to nil pointer dereference. This can be handled more gracefully by checking for the nil value.

If CRIU fails, the container engine (CRI-O/Podman) is not going to create a checkpoint archive. It might be better to show an appropriate error message instead of silently ignoring this.

> Secondly, this is related to the change in https://github.com/checkpoint-restore/checkpointctl/pull/87 - if we prune the process tree for a particular PID, the other PID nodes will not be present. Hence, for those PID's file descriptors, tree.FindByMeta() will return nil.

To enable process tree output for specified PID we can introduce a similar change in #87. For example, it might be more efficient to pass the value specified with `--pid` as a parameter to `addFdsToTree()` and use something like the following:
```Go
if pid != fd.PId {
	continue
}
```

Reply:
If you prefer, I can revert this change, and wait until this PR is merged. Since the change is directly related to #87, I can introduce this there after rebasing. What do you think?

Reply:
It would make sense to introduce this change in #87 after rebase.